### PR TITLE
fix!: london hardfork check logic in json-rpc apis

### PR DESCRIPTION
**Pull request by yihuang**
## Description

Closes: #1067
Solution:
- Add `BaseFee` grpc query handler in evm module which also checks the current london hardfork status.

the `eth_getBlockByNumber` rpc api is tested in cronos's related PR: https://github.com/crypto-org-chain/cronos/pull/446/files#diff-ece6d4f65047a3ec6f2294c4e4fc1eecf8a2b44928f478867399e69f194fc1beR128

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: rpc/ethereum/backend/backend.go [hunk 1]
```diff
@@ -991,13 +991,8 @@ func (e *EVMBackend) SuggestGasTipCap(baseFee *big.Int) (*big.Int, error) {
 // If the London hard fork is not activated at the current height, the query will
 // return nil.
 func (e *EVMBackend) BaseFee(height int64) (*big.Int, error) {
-	cfg := e.ChainConfig()
-	if !cfg.IsLondon(new(big.Int).SetInt64(height)) {
-		return nil, nil
-	}
-
-	// return BaseFee if London hard fork is activated and feemarket is not enabled
-	res, err := e.queryClient.FeeMarket.BaseFee(types.ContextWithHeight(height), &feemarkettypes.QueryBaseFeeRequest{})
```
Comment (fedekunze) [inline rpc/ethereum/backend/backend.go:1000]: should we remove the FeeMarket client from the JSON-RPC if it's not being used?

File: x/evm/keeper/grpc_query.go [hunk 2]
```diff
@@ -583,3 +583,20 @@ func (k *Keeper) traceTx(
 
 	return &result, txConfig.LogIndex + uint(len(res.Logs)), nil
 }
+
+// BaseFee implements the Query/BaseFee gRPC method
+func (k Keeper) BaseFee(c context.Context, _ *types.QueryBaseFeeRequest) (*types.QueryBaseFeeResponse, error) {
```
Comment (fedekunze) [inline x/evm/keeper/grpc_query.go:588]: can you create tests for the base fee query?

File: rpc/ethereum/backend/backend.go (hunk 1 above)
Comment (yihuang) [reply rpc/ethereum/backend/backend.go:1000]: there's a usage in `SuggestGasTipCap` to fetch feemarket parameters.

File: x/evm/keeper/grpc_query.go (hunk 2 above)
Comment (yihuang) [reply x/evm/keeper/grpc_query.go:588]: done

File: x/evm/keeper/keeper.go [hunk 3]
```diff
@@ -291,7 +291,7 @@ func (k *Keeper) GetBalance(ctx sdk.Context, addr common.Address) *big.Int {
 // - `nil`: london hardfork not enabled.
 // - `0`: london hardfork enabled but feemarket is not enabled.
 // - `n`: both london hardfork and feemarket are enabled.
-func (k Keeper) BaseFee(ctx sdk.Context, ethCfg *params.ChainConfig) *big.Int {
+func (k Keeper) GetBaseFee(ctx sdk.Context, ethCfg *params.ChainConfig) *big.Int {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/keeper.go:294]: yes, moved to api breaking section now.

File: x/evm/keeper/grpc_query_test.go [hunk 4]
```diff
@@ -899,3 +900,76 @@ func (suite *KeeperTestSuite) TestNonceInQuery() {
...
+				suite.Require().NotNil(res)
+				suite.Require().Equal(expRes, res, tc.name)
+				suite.Require().NoError(err)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+	suite.enableFeemarket = false
+	suite.enableLondonHF = true
```
Comment (fedekunze) [inline x/evm/keeper/grpc_query_test.go:974]: q: why do you need to change this after the tests?

**Review by fedekunze — APPROVED** [review]

File: x/evm/keeper/grpc_query_test.go (hunk 4 above)
Comment (yihuang) [reply x/evm/keeper/grpc_query_test.go:974]: To change it back to the default ones, so it doesn't affect the test cases executed after this one, without these lines, some other test case will fail.

File: x/evm/keeper/grpc_query_test.go [hunk 5]
```diff
@@ -899,3 +900,76 @@ func (suite *KeeperTestSuite) TestNonceInQuery() {
...
+			func() {
+				baseFee := sdk.OneInt().BigInt()
+				suite.app.FeeMarketKeeper.SetBaseFee(suite.ctx, baseFee)
+
+				expRes = &types.QueryBaseFeeResponse{}
+			},
+			true, true, false,
+		},
+		{
+			"pass - nil Base Fee when feemarket not activated",
```
Comment (yihuang) [inline x/evm/keeper/grpc_query_test.go:947]:
```suggestion
			"pass - zero Base Fee when feemarket not activated",
```

**Pull request merged**